File: static/js/sparsify.js
/**
 * Sparsify voxel clouds while preserving front (-Z) and side (-Y) projections.
 * Optimized for browser: O(voxels) memory, no full-cube scans in hot paths.
 */

import {
  applyEdgeWallWrap,
  boundaryFaces,
  pickYForSide,
  pickZForFront,
  yAtColumn,
  zPool,
} from "./carve_helpers.js";

function evenlyPick(indices, count) {
  if (count >= indices.length) return indices.slice();
  if (count <= 1) return [indices[Math.floor(indices.length / 2)]];
  const picked = new Set();
  for (let i = 0; i < count; i++) {
    const pos = Math.round((i * (indices.length - 1)) / Math.max(1, count - 1));
    picked.add(indices[pos]);
  }
  return [...picked].sort((a, b) => a - b);
}

function pickBestY(validY, x, z, points) {
  if (points.length === 0) return validY[Math.floor(validY.length / 2)];
  let bestY = validY[0];
  let bestD = -1;
  for (const y of validY) {
    let minD = Infinity;
    for (let i = 0; i < points.length; i++) {
      const py = points[i][0];
      const px = points[i][1];
      const pz = points[i][2];
      const d = (py - y) ** 2 + (px - x) ** 2 + (pz - z) ** 2;
      if (d < minD) minD = d;
    }
    if (minD > bestD) {
      bestD = minD;
      bestY = y;
    }
  }
  return bestY;
}

function repairCoverageMasks(
  selected,
  reqFront,
  reqSide,
  front,
  side,
  size,
  colCount,
  rowCount,
  points,
  depthFaceBridge,
  edgeStripFill,
  edgeWallWrap
) {
  const idx = (y, x, z) => y * size * size + x * size + z;
  let fallbackYFaces = null;
  let fallbackZFaces = null;
  if (edgeStripFill || edgeWallWrap) {
    const faces = boundaryFaces(front, side, size);
    fallbackYFaces = faces.yFaces;
    fallbackZFaces = faces.zFaces;
  }

  const add = (y, x, z) => {
    if (selected[idx(y, x, z)]) return;
    selected[idx(y, x, z)] = 1;
    colCount[y * size + x]++;
    rowCount[z * size + x]++;
    points.push([y, x, z]);
  };

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (!reqFront[y * size + x] || colCount[y * size + x] > 0) continue;
      const z = pickZForFront(y, x, side, size, depthFaceBridge, fallbackZFaces);
      if (z != null) add(y, x, z);
    }
  }

  for (let z = 0; z < size; z++) {
    for (let x = 0; x < size; x++) {
      if (!reqSide[z * size + x] || rowCount[z * size + x] > 0) continue;
      const y = pickYForSide(z, x, front, size, depthFaceBridge, fallbackYFaces);
      if (y != null) add(y, x, z);
    }
  }
}

export function sparsifyUniform(
  voxels,
  size,
  density = 0.4,
  uniformStrength = 0.6,
  targetFront = null,
  targetSide = null,
  depthFaceBridge = true,
  edgeStripFill = true,
  edgeWallWrap = true
) {
  density = Math.max(0.05, Math.min(1, density));
  uniformStrength = Math.max(0, Math.min(1, uniformStrength));

  const idx = (y, x, z) => y * size * size + x * size + z;
  let total = 0;
  for (let i = 0; i < voxels.length; i++) total += voxels[i];

  const reqFront = targetFront || new Uint8Array(size * size);
  const reqSide = targetSide || new Uint8Array(size * size);
  if (!targetFront) {
    for (let y = 0; y < size; y++) {
      for (let x = 0; x < size; x++) {
        for (let z = 0; z < size; z++) {
          if (voxels[idx(y, x, z)]) {
            reqFront[y * size + x] = 1;
            break;
          }
        }
      }
    }
  }
  if (!targetSide) {
    for (let z = 0; z < size; z++) {
      for (let x = 0; x < size; x++) {
        for (let y = 0; y < size; y++) {
          if (voxels[idx(y, x, z)]) {
            reqSide[z * size + x] = 1;
            break;
          }
        }
      }
    }
  }

  const front = reqFront;
  const side = reqSide;

  if (density >= 0.999) {
    const out = new Uint8Array(voxels);
    const colCount = new Uint16Array(size * size);
    const rowCount = new Uint16Array(size * size);
    const points = [];
    for (let y = 0; y < size; y++) {
      for (let x = 0; x < size; x++) {
        for (let z = 0; z < size; z++) {
          if (out[idx(y, x, z)]) {
            colCount[y * size + x]++;
            rowCount[z * size + x]++;
            points.push([y, x, z]);
          }
        }
      }
    }
    repairCoverageMasks(
      out,
      reqFront,
      reqSide,
      front,
      side,
      size,
      colCount,
      rowCount,
      points,
      depthFaceBridge,
      edgeStripFill,
      edgeWallWrap
    );
    if (edgeWallWrap) applyEdgeWallWrap(out, front, side, size);
    return out;
  }

  if (total === 0 && !reqFront.some?.((v) => v) && !reqSide.some?.((v) => v)) {
    let any = false;
    for (let i = 0; i < reqFront.length; i++) if (reqFront[i]) any = true;
    for (let i = 0; i < reqSide.length; i++) if (reqSide[i]) any = true;
    if (!any) return voxels;
  }

  const selected = new Uint8Array(voxels.length);
  const colCount = new Uint16Array(size * size);
  const rowCount = new Uint16Array(size * size);
  const points = [];

  const addPoint = (y, x, z) => {
    if (selected[idx(y, x, z)]) return;
    selected[idx(y, x, z)] = 1;
    colCount[y * size + x]++;
    rowCount[z * size + x]++;
    points.push([y, x, z]);
  };

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      if (!reqFront[y * size + x]) continue;
      let zList = zPool(side, size, x, depthFaceBridge);
      if (!zList.length) continue;
      const n = Math.max(1, Math.ceil(zList.length * density));
      const keep =
        uniformStrength > 0.2
          ? evenlyPick(zList, n)
          : zList.filter((_, i) => i % Math.max(1, Math.floor(zList.length / n)) === 0).slice(0, n);
      for (const z of keep) addPoint(y, x, z);
    }
  }

  for (let z = 0; z < size; z++) {
    for (let x = 0; x < size; x++) {
      if (!reqSide[z * size + x] || rowCount[z * size + x] > 0) continue;
      let yList = yAtColumn(front, size, x);
      if (!yList.length) continue;
      const y =
        uniformStrength > 0.2 && points.length < 8000
          ? pickBestY(yList, x, z, points)
          : yList[(z * yList.length) / size % yList.length | 0];
      addPoint(y, x, z);
    }
  }

  // Phase 3: 3D grid dedup (iterate points list only)
  if (uniformStrength > 0.15 && points.length > 0) {
    const cell = Math.max(2, Math.round(size * (0.06 + 0.22 * uniformStrength * (1 - density * 0.5))));
    const buckets = new Map();
    for (const [y, x, z] of points) {
      const ky = Math.floor(y / cell);
      const kx = Math.floor(x / cell);
      const kz = Math.floor(z / cell);
      const k = `${ky},${kx},${kz}`;
      const cy = (ky + 0.5) * cell;
      const cx = (kx + 0.5) * cell;
      const cz = (kz + 0.5) * cell;
      const dist = (y - cy) ** 2 + (x - cx) ** 2 + (z - cz) ** 2;
      const prev = buckets.get(k);
      if (!prev || dist < prev.dist) buckets.set(k, { dist, y, x, z });
    }
    selected.fill(0);
    colCount.fill(0);
    rowCount.fill(0);
    points.length = 0;
    for (const { y, x, z } of buckets.values()) addPoint(y, x, z);
    repairCoverageMasks(
      selected,
      reqFront,
      reqSide,
      front,
      side,
      size,
      colCount,
      rowCount,
      points,
      depthFaceBridge,
      edgeStripFill,
      edgeWallWrap
    );
  }

  // Phase 4: batch greedy removal (max iterations capped)
  let reqFrontCount = 0;
  let reqSideCount = 0;
  for (let i = 0; i < reqFront.length; i++) reqFrontCount += reqFront[i];
  for (let i = 0; i < reqSide.length; i++) reqSideCount += reqSide[i];

  let target = Math.max(Math.floor((reqFrontCount + reqSideCount) / 2), Math.floor(total * density));
  target = Math.max(target, Math.floor(reqFrontCount * 0.3));

  const maxPasses = 64;
  let pass = 0;
  while (points.length > target && pass < maxPasses) {
    pass++;
    const batch = Math.min(256, points.length - target);
    const candidates = [];
    for (let i = 0; i < points.length; i++) {
      const [y, x, z] = points[i];
      const col = colCount[y * size + x];
      const row = rowCount[z * size + x];
      if (col <= 1 && row <= 1) continue;
      candidates.push({ score: col + row, i, y, x, z });
    }
    if (!candidates.length) break;
    candidates.sort((a, b) => b.score - a.score);
    const removeIdx = new Set();
    for (let k = 0; k < Math.min(batch, candidates.length); k++) {
      removeIdx.add(candidates[k].i);
    }
    const next = [];
    for (let i = 0; i < points.length; i++) {
      if (removeIdx.has(i)) {
        const [y, x, z] = points[i];
        selected[idx(y, x, z)] = 0;
        colCount[y * size + x]--;
        rowCount[z * size + x]--;
      } else {
        next.push(points[i]);
      }
    }
    points.length = 0;
    points.push(...next);
    repairCoverageMasks(
      selected,
      reqFront,
      reqSide,
      front,
      side,
      size,
      colCount,
      rowCount,
      points,
      depthFaceBridge,
      edgeStripFill,
      edgeWallWrap
    );
  }

  if (edgeWallWrap) applyEdgeWallWrap(selected, front, side, size);
  return selected;
}

export function voxelsToPoints(selected, size) {
  const idx = (y, x, z) => y * size * size + x * size + z;
  const points = [];
  const projectionFront = Array.from({ length: size }, () => Array(size).fill(0));
  const projectionSide = Array.from({ length: size }, () => Array(size).fill(0));

  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      for (let z = 0; z < size; z++) {
        if (!selected[idx(y, x, z)]) continue;
        points.push([x, y, z]);
        projectionFront[y][x] = 1;
        projectionSide[z][x] = 1;
      }
    }
  }
  return { points, projectionFront, projectionSide };
}

File: sparsify.py
"""Sparsify voxel clouds while preserving dual-view projections."""

from __future__ import annotations

import numpy as np

from carve_helpers import (
    apply_edge_wall_wrap,
    boundary_faces,
    pick_y_for_side,
    pick_z_for_front,
    y_at_column,
    z_at_column,
)


def _evenly_pick(indices: np.ndarray, count: int) -> np.ndarray:
    if count >= indices.size:
        return indices
    if count <= 1:
        return indices[indices.size // 2 : indices.size // 2 + 1]
    pos = np.linspace(0, indices.size - 1, count)
    return indices[np.unique(np.round(pos).astype(int))]


def _repair_coverage(
    selected: np.ndarray,
    target_front: np.ndarray,
    target_side: np.ndarray,
    front: np.ndarray,
    side: np.ndarray,
    *,
    depth_face_bridge: bool = True,
    edge_strip_fill: bool = True,
    edge_wall_wrap: bool = True,
) -> None:
    size = front.shape[0]
    fallback_y_faces: np.ndarray | None = None
    fallback_z_faces: np.ndarray | None = None
    if edge_strip_fill or edge_wall_wrap:
        fallback_y_faces, fallback_z_faces = boundary_faces(front, side)

    for y in range(size):
        for x in range(size):
            if not target_front[y, x] or np.any(selected[y, x, :]):
                continue
            z = pick_z_for_front(
                y,
                x,
                side,
                size,
                depth_face_bridge,
                fallback_z_faces=fallback_z_faces,
            )
            if z is not None:
                selected[y, x, z] = True

    for z in range(size):
        for x in range(size):
            if not target_side[z, x] or np.any(selected[:, x, z]):
                continue
            y = pick_y_for_side(
                z,
                x,
                front,
                size,
                depth_face_bridge,
                fallback_y_faces=fallback_y_faces,
            )
            if y is not None:
                selected[y, x, z] = True


def _z_pool(side: np.ndarray, x: int, depth_face_bridge: bool) -> np.ndarray:
    strict = z_at_column(side, x)
    if strict.size == 0:
        return strict
    z_lo, z_hi = int(strict[0]), int(strict[-1])
    if depth_face_bridge and z_hi > z_lo + 1:
        return np.unique(np.concatenate([strict, np.array([z_lo, z_hi], dtype=int)]))
    return strict


def _pick_best_y(valid_y: np.ndarray, x: int, z: int, existing: np.ndarray) -> int:
    if existing.size == 0:
        return int(valid_y[valid_y.size // 2])
    best_y = int(valid_y[0])
    best_d = -1.0
    for y in valid_y:
        d = float(np.min((existing[:, 0] - y) ** 2 + (existing[:, 1] - x) ** 2 + (existing[:, 2] - z) ** 2))
        if d > best_d:
            best_d = d
            best_y = int(y)
    return best_y


def sparsify_uniform(
    voxels: np.ndarray,
    density: float = 0.4,
    uniform_strength: float = 0.6,
    target_front: np.ndarray | None = None,
    target_side: np.ndarray | None = None,
    depth_face_bridge: bool = True,
    edge_strip_fill: bool = True,
    edge_wall_wrap: bool = True,
) -> np.ndarray:
    density = float(np.clip(density, 0.05, 1.0))
    uniform_strength = float(np.clip(uniform_strength, 0.0, 1.0))

    size = voxels.shape[0]
    req_front = target_front if target_front is not None else np.any(voxels, axis=2)
    req_side = target_side if target_side is not None else np.any(voxels, axis=0).T
    front = req_front
    side = req_side

    if density >= 0.999:
        out = voxels.copy()
        _repair_coverage(
            out,
            req_front,
            req_side,
            front,
            side,
            depth_face_bridge=depth_face_bridge,
            edge_strip_fill=edge_strip_fill,
            edge_wall_wrap=edge_wall_wrap,
        )
        if edge_wall_wrap:
            apply_edge_wall_wrap(out, front, side)
        return out

    total = int(voxels.sum())
    selected = np.zeros_like(voxels)

    for y in range(size):
        for x in range(size):
            if not req_front[y, x]:
                continue
            z_cands = _z_pool(side, x, depth_face_bridge)
            if z_cands.size == 0:
                continue
            n = max(1, int(np.ceil(z_cands.size * density)))
            if uniform_strength > 0.2:
                keep = _evenly_pick(z_cands, n)
            else:
                step = max(1, z_cands.size // n)
                keep = z_cands[::step][:n]
            selected[y, x, keep] = True

    existing = np.argwhere(selected)
    for z in range(size):
        for x in range(size):
            if not req_side[z, x] or np.any(selected[:, x, z]):
                continue
            y_cands = y_at_column(front, x)
            if y_cands.size == 0:
                continue
            if uniform_strength > 0.2 and existing.size < 8000:
                y = _pick_best_y(y_cands, x, z, existing)
            else:
                y = int(y_cands[(z * y_cands.size) // max(1, size) % y_cands.size])
            selected[y, x, z] = True
            existing = np.argwhere(selected)

    if uniform_strength > 0.15:
        cell = max(2, int(round(size * (0.06 + 0.22 * uniform_strength * (1.0 - density * 0.5)))))
        buckets: dict[tuple[int, int, int], tuple[float, int, int, int]] = {}
        for y, x, z in np.argwhere(selected):
            key = (y // cell, x // cell, z // cell)
            cy = (key[0] + 0.5) * cell
            cx = (key[1] + 0.5) * cell
            cz = (key[2] + 0.5) * cell
            dist = (y - cy) ** 2 + (x - cx) ** 2 + (z - cz) ** 2
            if key not in buckets or dist < buckets[key][0]:
                buckets[key] = (dist, int(y), int(x), int(z))
        selected.fill(False)
        for _, y, x, z in buckets.values():
            selected[y, x, z] = True
        _repair_coverage(
            selected,
            req_front,
            req_side,
            front,
            side,
            depth_face_bridge=depth_face_bridge,
            edge_strip_fill=edge_strip_fill,
            edge_wall_wrap=edge_wall_wrap,
        )

    target = max(int(req_front.sum() + req_side.sum()) // 2, int(total * density))
    target = max(target, int(req_front.sum() * 0.3))

    coords = np.argwhere(selected)
    max_passes = 64
    for _ in range(max_passes):
        if coords.shape[0] <= target:
            break
        col = selected.sum(axis=2)
        row = selected.sum(axis=0)
        removable = []
        for y, x, z in coords:
            if col[y, x] <= 1 and row[x, z] <= 1:
                continue
            removable.append((int(col[y, x] + row[x, z]), int(y), int(x), int(z)))
        if not removable:
            break
        removable.sort(reverse=True)
        batch = min(256, coords.shape[0] - target, len(removable))
        for _, y, x, z in removable[:batch]:
            selected[y, x, z] = False
        _repair_coverage(
            selected,
            req_front,
            req_side,
            front,
            side,
            depth_face_bridge=depth_face_bridge,
            edge_strip_fill=edge_strip_fill,
            edge_wall_wrap=edge_wall_wrap,
        )
        coords = np.argwhere(selected)

    _repair_coverage(
        selected,
        req_front,
        req_side,
        front,
        side,
        depth_face_bridge=depth_face_bridge,
        edge_strip_fill=edge_strip_fill,
        edge_wall_wrap=edge_wall_wrap,
    )
    if edge_wall_wrap:
        apply_edge_wall_wrap(selected, front, side)
    return selected
